source/scene/forwardscenerenderpass.ts

```diff
@@ -0,0 +1,180 @@
+
+import { assert } from '../auxiliaries';
+import { GLfloat2 } from '../tuples';
+
+import { Camera } from '../camera';
+import { ChangeLookup } from '../changelookup';
+import { Context } from '../context';
+import { Framebuffer } from '../framebuffer';
+import { Initializable } from '../initializable';
+
+import { SceneRenderPass } from './scenerenderpass';
+import { SceneNode } from './scenenode';
+import { mat4 } from 'gl-matrix';
+import { GeometryComponent } from './geometrycomponent';
+
+
+/**
+ * @todo add description
+ */
+export class ForwardSceneRenderPass extends SceneRenderPass {
+
+    /**
+     * Alterable auxiliary object for tracking changes on render pass inputs and lazy updates.
+     */
+    protected readonly _altered = Object.assign(new ChangeLookup(), {
+        any: false,
+        camera: false,
+    });
+
+    /** @see {@link target} */
+    protected _target: Framebuffer;
+
+    /** @see {@link camera} */
+    protected _camera: Camera;
+
+    /** @see {@link ndcOffset} */
+    protected _ndcOffset: GLfloat2;
+
+
+    /**
+     * Creates a ...
+     * @param context - @todo comment
+     */
+    constructor(context: Context) {
+        super();
+        this._context = context;
+
+        /** @todo */
+    }
+
+
+    @Initializable.initialize()
+    initialize(): boolean {
+        // const gl = this._context.gl;
+
+        /** @todo create shaders, programs, fbos, etc. - checkout label render pass for example */
+
+        return true;
+    }
+
+    @Initializable.uninitialize()
+    uninitialize(): void {
+
+        /** @todo create shaders, programs, fbos, etc. - checkout label render pass for example */
+
+    }
+
+
+    /**
+     * @todo comment
+     */
+    prepare(): void {
+
+        /** @todo prepare for immediate (probably multiple) frame invocations (mostly useful in multi-frame sampling) */
+
+    }
+
+
+    /**
+     * @param override - If enabled, everything will be updated, regardless of tracked alterations.
+     */
+    @Initializable.assert_initialized()
+    update(override: boolean = false): void {
+        // const gl = this._context.gl;
+
+        /** @todo  checkout label render pass update for reference */
+
+        // this._scene ... from SceneNodeRenderer ...
+
+    }
+
+    /**
+     * @todo comment
+     */
+    @Initializable.assert_initialized()
+    frame(): void {
+        assert(this._target && this._target.valid, `valid target expected`);
+        if (this._scene === undefined) {
+            return;
+        }
+
+        const gl = this._context.gl;
+
+        const size = this._target.size;
+        gl.viewport(0, 0, size[0], size[1]);
```

Comment:
The viewport is updated here...I don't think the renderer above should assume its control over the OpenGL viewport.